### PR TITLE
Fixes for PLUGIN-1533 PLUGIN-1469 PLUGIN-1498

### PR DESCRIPTION
**Pull request by sgarg-CS**
[PLUGIN-1469] (https://cdap.atlassian.net/browse/PLUGIN-1469)
[PLUGIN-1498](https://cdap.atlassian.net/browse/PLUGIN-1498) 
[PLUGIN-1533](https://cdap.atlassian.net/browse/PLUGIN-1533) 

[PLUGIN-1469]: https://cdap.atlassian.net/browse/PLUGIN-1469?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLUGIN-1498]: https://cdap.atlassian.net/browse/PLUGIN-1498?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLUGIN-1533]: https://cdap.atlassian.net/browse/PLUGIN-1533?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/Salesforce-batchsink.md [hunk 1]
```diff
@@ -81,6 +81,13 @@ Update - updates existing records based on Id field.
 **Upsert External ID Field:** External id field name. It is used only if operation is upsert.
 The field specified can be either 'Id' or any customly created field, which has external id attribute set.
 
+**Concurrency Mode:** The concurrency mode for the job. The valid values are:<br>
```
Comment (greeshmaswaminathan) [inline docs/Salesforce-batchsink.md]: nit: "The concurrency mode for the job." -> "The concurrency mode for the bulk job."

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 2]
```diff
@@ -85,6 +87,17 @@ public class SalesforceSinkConfig extends BaseSalesforceConfig {
   @Macro
   private String externalIdField;
 
+  @Name(PROPERTY_CONCURRENCY_MODE)
+  @Description("The concurrency mode for the job. The valid values are: \n" +
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: nit: bulk job.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 3]
```diff
@@ -148,6 +162,19 @@ public String getExternalIdField() {
     return externalIdField;
   }
 
+  public String getConcurrencyMode() {
+    return concurrencyMode;
+  }
+
+  public ConcurrencyMode getConcurrencyModeEnum() {
+    try {
+      return ConcurrencyMode.valueOf(concurrencyMode);
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Do we need a null check and default value setting here? Test how pipeline upgrade works when we are upgrading from 1.4.3 to 1.4.4. version of the plugin.

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java [hunk 4]
```diff
@@ -106,8 +107,8 @@ private static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String qu
     throws AsyncApiException, IOException {
 
     SObjectDescriptor sObjectDescriptor = SObjectDescriptor.fromQuery(query);
-    JobInfo job = SalesforceBulkUtil.createJob(bulkConnection, sObjectDescriptor.getName(),
-            getOperationEnum(operation), null);
+    JobInfo job = SalesforceBulkUtil.createJob(bulkConnection, sObjectDescriptor.getName(), getOperationEnum(operation),
+                                               null, ConcurrencyMode.Parallel);
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java:111]: What is the use case here? Should the concurrency mode be passed here from the config ?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 5]
```diff
@@ -157,6 +157,7 @@ public void createSalesforceListener() throws Exception {
 
         boolean success = message.isSuccessful();
         if (!success) {
+          LOG.info(String.format("Error in meta connect, message: %s", message));
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Change to LOG.error

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 6]
```diff
@@ -176,7 +177,7 @@ public void createSalesforceListener() throws Exception {
             }
           } else {
             throw new RuntimeException(String.format("Error in meta connect, errorMessage: %s Advice: %s", error,
-                                                     advice));
+                    advice));
```
Comment (greeshmaswaminathan) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Formatting looks incorrect here.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Corrected it. e130c2d587656a458efc1c7b64df2f795710543c

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java (hunk 5 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Changed e130c2d587656a458efc1c7b64df2f795710543c

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java (hunk 4 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java:111]: The Salesforce Batch Source as well as Multi-source plugins also use the bulk job API and call the same `createJob()` method. 
No changes have been made as there aren't any lock contention issues observed in the Source plugins.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Returning the default value: `ConcurrencyMode.Parallel` if 

- the value is null or
- this enum type has no constant with the specified name

Tested with pipeline upgrade and it returns 200 OK. No issues were observed.

e130c2d587656a458efc1c7b64df2f795710543c

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 2 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: done e130c2d587656a458efc1c7b64df2f795710543c

File: docs/Salesforce-batchsink.md (hunk 1 above)
Comment (sgarg-CS) [reply docs/Salesforce-batchsink.md]: done e130c2d587656a458efc1c7b64df2f795710543c

**Comment by sgarg-CS** [comment]
Also, added a new method to renew the session if the bulk connection resets.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Please add a log in the exception block, so we know that there was an issue with the concurrency mode.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Added logger in the exception catch block. b1f4d0dd2ca351b7b9e6dbc0ad8b5b518a9402c2

**Comment by sgarg-CS** [comment]
All commits have been squashed.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 3 above)
Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Please change the log to info, and also mention that parallel mode is being used.

Comment (greeshmaswaminathan) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Add to the log message that default concurrency mode is used instead. The intention here is to inform that we encountered an issue and took an action.

Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java:173]: Updated the log message.

**Comment by sgarg-CS** [comment]
Again squashed all commits.

**Review by greeshmaswaminathan — APPROVED** [review]

**Pull request merged**